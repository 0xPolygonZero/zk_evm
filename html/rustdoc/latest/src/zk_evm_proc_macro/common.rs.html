<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Source of the Rust file `proc_macro/src/common.rs`."><title>common.rs - source</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-492a78a4a87dcc01.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="zk_evm_proc_macro" data-themes="" data-resource-suffix="" data-rustdoc-version="1.82.0-nightly (1f12b9b0f 2024-08-27)" data-channel="nightly" data-search-js="search-acd2e111c9cb27e3.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../../static.files/src-script-e66d777a5a92e9b2.js"></script><script defer src="../../src-files.js"></script><script defer src="../../static.files/main-cb0df477c2d67d00.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-3b12f09e550e0385.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/katex.min.css"
    integrity="sha384-zB1R0rpPzHqg7Kpt0Aljp8JPLqbXI3bhnPWROx27a9N0Ll6ZP/+DiW/UqRcLbRjq" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/katex.min.js"
    integrity="sha384-y23I5Q6l+B6vatafAwxRu/0oK/79VlbSz7Q9aiSZUvyWYIYsd+qj+o24G5ZU2zJz"
    crossorigin="anonymous"></script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/contrib/auto-render.min.js"
    integrity="sha384-kWPLUVMOks5AQFrykwIup5lo0m3iMkkHrD0uJ4H5cjeGihAutqP0yW0J6dpFiVkI"
    crossorigin="anonymous"></script>
<script>
    document.addEventListener("DOMContentLoaded", function () {
        renderMathInElement(document.body, {
            fleqn: false,
            macros: {
                "\\F": "\\mathbb{F}",
                "\\G": "\\mathbb{G}",
                "\\O": "\\mathcal{O}",
                "\\(": "\\left(",
                "\\)": "\\right)",
                "\\norm": "\\left\\vert #1 \\right\\vert",
                "\\set": "\\mathcal{ #1 }",
            },
            delimiters: [
                { left: "$$", right: "$$", display: true },
                { left: "\\(", right: "\\)", display: false },
                { left: "$", right: "$", display: false },
                { left: "\\[", right: "\\]", display: true }
            ]
        });
    });
</script>
</head><body class="rustdoc src"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="src-sidebar-title"><h2>Files</h2></div></nav><div class="sidebar-resizer"></div><main><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="example-wrap"><div data-nosnippet><pre class="src-line-numbers"><a href="#1" id="1">1</a>
<a href="#2" id="2">2</a>
<a href="#3" id="3">3</a>
<a href="#4" id="4">4</a>
<a href="#5" id="5">5</a>
<a href="#6" id="6">6</a>
<a href="#7" id="7">7</a>
<a href="#8" id="8">8</a>
<a href="#9" id="9">9</a>
<a href="#10" id="10">10</a>
<a href="#11" id="11">11</a>
<a href="#12" id="12">12</a>
<a href="#13" id="13">13</a>
<a href="#14" id="14">14</a>
<a href="#15" id="15">15</a>
<a href="#16" id="16">16</a>
<a href="#17" id="17">17</a>
<a href="#18" id="18">18</a>
<a href="#19" id="19">19</a>
<a href="#20" id="20">20</a>
<a href="#21" id="21">21</a>
<a href="#22" id="22">22</a>
<a href="#23" id="23">23</a>
<a href="#24" id="24">24</a>
<a href="#25" id="25">25</a>
<a href="#26" id="26">26</a>
<a href="#27" id="27">27</a>
<a href="#28" id="28">28</a>
<a href="#29" id="29">29</a>
<a href="#30" id="30">30</a>
<a href="#31" id="31">31</a>
<a href="#32" id="32">32</a>
<a href="#33" id="33">33</a>
<a href="#34" id="34">34</a>
<a href="#35" id="35">35</a>
<a href="#36" id="36">36</a>
<a href="#37" id="37">37</a>
<a href="#38" id="38">38</a>
<a href="#39" id="39">39</a>
<a href="#40" id="40">40</a>
<a href="#41" id="41">41</a>
</pre></div><pre class="rust"><code><span class="kw">use </span>syn::punctuated::Punctuated;
<span class="kw">use </span>syn::{token, Attribute, Meta};

<span class="doccomment">/// Prefixes an error message and generates a `syn::Error` from the message.
</span><span class="macro">macro_rules!</span> span_err {
    (<span class="macro-nonterminal">$ast</span>:expr, <span class="macro-nonterminal">$msg</span>:literal $(,)<span class="question-mark">?</span>) =&gt; {
        ::syn::Error::new_spanned(<span class="macro-nonterminal">$ast</span>, <span class="macro">::core::concat!</span>(<span class="string">"zk_evm_proc_macro error: "</span>, <span class="macro-nonterminal">$msg</span>))
    };
}
<span class="kw">pub</span>(<span class="kw">crate</span>) <span class="kw">use </span>span_err;

<span class="doccomment">/// Checks the condition and returns early with a prefixed error message if
/// false.
</span><span class="macro">macro_rules!</span> ensure {
    (<span class="macro-nonterminal">$cond</span>:expr, <span class="macro-nonterminal">$ast</span>:expr, <span class="macro-nonterminal">$msg</span>:literal $(,)<span class="question-mark">?</span>) =&gt; {
        <span class="kw">if </span>!<span class="macro-nonterminal">$cond </span>{
            <span class="kw">return </span><span class="prelude-val">Err</span>(<span class="macro-nonterminal">$</span><span class="macro">crate::common::span_err!</span>(<span class="macro-nonterminal">$ast</span>, <span class="macro-nonterminal">$msg</span>));
        }
    };
}
<span class="kw">pub</span>(<span class="kw">crate</span>) <span class="kw">use </span>ensure;

<span class="doccomment">/// Parses the `Meta` of a `repr` attribute and returns true if one of the
/// elements is "C".
</span><span class="kw">fn </span>is_meta_c(outer: <span class="kw-2">&amp;</span>Meta) -&gt; bool {
    <span class="kw">if let </span>Meta::List(inner) = outer {
        <span class="kw">let </span>parsed: Punctuated&lt;Meta, token::Comma&gt; = inner
            .parse_args_with(Punctuated::parse_terminated)
            .unwrap_or_default();
        parsed.iter().any(|meta| meta.path().is_ident(<span class="string">"C"</span>))
    } <span class="kw">else </span>{
        <span class="bool-val">false
    </span>}
}

<span class="doccomment">/// Returns true if `#[repr(C)]` is contained in the attributes.
</span><span class="kw">pub</span>(<span class="kw">crate</span>) <span class="kw">fn </span>is_repr_c&lt;<span class="lifetime">'a</span>&gt;(attrs: <span class="kw">impl </span>IntoIterator&lt;Item = <span class="kw-2">&amp;</span><span class="lifetime">'a </span>Attribute&gt;) -&gt; bool {
    attrs
        .into_iter()
        .any(|attr| attr.path().is_ident(<span class="string">"repr"</span>) &amp;&amp; is_meta_c(<span class="kw-2">&amp;</span>attr.meta))
}
</code></pre></div></section></main></body></html>