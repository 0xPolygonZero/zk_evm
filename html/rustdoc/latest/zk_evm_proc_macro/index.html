<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This library provides two convenient derive macros for interpreting arrays of field elements as structs representing an AIR."><title>zk_evm_proc_macro - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-492a78a4a87dcc01.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="zk_evm_proc_macro" data-themes="" data-resource-suffix="" data-rustdoc-version="1.82.0-nightly (1f12b9b0f 2024-08-27)" data-channel="nightly" data-search-js="search-acd2e111c9cb27e3.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-cb0df477c2d67d00.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-3b12f09e550e0385.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/katex.min.css"
    integrity="sha384-zB1R0rpPzHqg7Kpt0Aljp8JPLqbXI3bhnPWROx27a9N0Ll6ZP/+DiW/UqRcLbRjq" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/katex.min.js"
    integrity="sha384-y23I5Q6l+B6vatafAwxRu/0oK/79VlbSz7Q9aiSZUvyWYIYsd+qj+o24G5ZU2zJz"
    crossorigin="anonymous"></script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/contrib/auto-render.min.js"
    integrity="sha384-kWPLUVMOks5AQFrykwIup5lo0m3iMkkHrD0uJ4H5cjeGihAutqP0yW0J6dpFiVkI"
    crossorigin="anonymous"></script>
<script>
    document.addEventListener("DOMContentLoaded", function () {
        renderMathInElement(document.body, {
            fleqn: false,
            macros: {
                "\\F": "\\mathbb{F}",
                "\\G": "\\mathbb{G}",
                "\\O": "\\mathcal{O}",
                "\\(": "\\left(",
                "\\)": "\\right)",
                "\\norm": "\\left\\vert #1 \\right\\vert",
                "\\set": "\\mathcal{ #1 }",
            },
            delimiters: [
                { left: "$$", right: "$$", display: true },
                { left: "\\(", right: "\\)", display: false },
                { left: "$", right: "$", display: false },
                { left: "\\[", right: "\\]", display: true }
            ]
        });
    });
</script>
</head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../zk_evm_proc_macro/index.html">zk_<wbr>evm_<wbr>proc_<wbr>macro</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#derives">Derive Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">zk_evm_proc_macro</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/zk_evm_proc_macro/lib.rs.html#1-47">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>This library provides two convenient derive macros for interpreting arrays
of field elements as structs representing an AIR.</p>
<p>Deriving <a href="derive.Columns.html" title="derive zk_evm_proc_macro::Columns"><code>Columns</code></a> on a struct <code>Struct&lt;T&gt;</code> implements the following
conversion traits between <code>Struct&lt;T&gt;</code> and arrays <code>[T; N]</code> where <code>N</code> is the
number of fields in the struct: <a href="https://doc.rust-lang.org/nightly/core/borrow/trait.Borrow.html" title="trait core::borrow::Borrow"><code>Borrow</code></a>, <a href="https://doc.rust-lang.org/nightly/core/borrow/trait.BorrowMut.html" title="trait core::borrow::BorrowMut"><code>BorrowMut</code></a>, and <a href="https://doc.rust-lang.org/nightly/core/convert/trait.From.html" title="trait core::convert::From"><code>From</code></a>.
Additionally, the traits <a href="https://doc.rust-lang.org/nightly/core/ops/index/trait.Index.html" title="trait core::ops::index::Index"><code>Index</code></a>, <a href="https://doc.rust-lang.org/nightly/core/ops/index/trait.IndexMut.html" title="trait core::ops::index::IndexMut"><code>IndexMut</code></a>, and <a href="https://doc.rust-lang.org/nightly/core/default/trait.Default.html" title="trait core::default::Default"><code>Default</code></a> are
implemented for <code>Struct&lt;T&gt;</code>.</p>
<p>Deriving <a href="derive.DerefColumns.html" title="derive zk_evm_proc_macro::DerefColumns"><code>DerefColumns</code></a> for a struct generic over <code>T</code> implements <a href="https://doc.rust-lang.org/nightly/core/ops/deref/trait.Deref.html" title="trait core::ops::deref::Deref"><code>Deref</code></a>
and <a href="https://doc.rust-lang.org/nightly/core/ops/deref/trait.DerefMut.html" title="trait core::ops::deref::DerefMut"><code>DerefMut</code></a> with target <code>[T; N]</code> where <code>N</code> is the number of fields in
the struct.</p>
<p>These implementations employ unsafe code and place a burden on the user to
ensure their safe usage. Please see the respective macro implementations to
understand the conditions that should be upheld by any struct deriving
<a href="derive.Columns.html" title="derive zk_evm_proc_macro::Columns"><code>Columns</code></a> or <a href="derive.DerefColumns.html" title="derive zk_evm_proc_macro::DerefColumns"><code>DerefColumns</code></a>. In short, the struct must be <code>#[repr(C)]</code>
and all fields must be one of <code>T</code>, <code>[T; M]</code>, or a type with the same layout
as <code>[T; M]</code>.</p>
</div></details><h2 id="derives" class="section-header">Derive Macros<a href="#derives" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="derive" href="derive.Columns.html" title="derive zk_evm_proc_macro::Columns">Columns</a></div></li><li><div class="item-name"><a class="derive" href="derive.DerefColumns.html" title="derive zk_evm_proc_macro::DerefColumns">Deref<wbr>Columns</a></div></li></ul></section></div></main></body></html>