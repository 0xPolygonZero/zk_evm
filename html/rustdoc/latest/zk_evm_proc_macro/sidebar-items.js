window.SIDEBAR_ITEMS = {"derive":["Columns","DerefColumns"]};