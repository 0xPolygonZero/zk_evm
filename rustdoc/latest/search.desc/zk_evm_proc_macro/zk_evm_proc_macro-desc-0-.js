searchState.loadedDescShard("zk_evm_proc_macro", 0, "This library provides two convenient derive macros for â€¦")