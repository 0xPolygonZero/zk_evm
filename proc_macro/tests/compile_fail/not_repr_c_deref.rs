use zk_evm_proc_macro::DerefColumns;

#[derive(DerefColumns)]
struct Columns<T> {
    a: T,
    b: [T; 3],
}

fn main() {}
