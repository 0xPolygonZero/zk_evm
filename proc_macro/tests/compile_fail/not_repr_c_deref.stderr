error: zk_evm_proc_macro error: column struct must be `#[repr(C)]`
 --> tests/compile_fail/not_repr_c_deref.rs:4:1
  |
4 | / struct Columns<T> {
5 | |     a: T,
6 | |     b: [T; 3],
7 | | }
  | |_^
